### PR TITLE
Scottx611x/django docker engine integration

### PR DESCRIPTION
**Pull request by scottx611x**
Starting point for completing #1672

- Ability to generate Visualization ToolDefinitions from a properly written ToolAnnotation
- Generic endpoint to launch tools from (for  visualizations only atm)
- Integration of https://github.com/refinery-platform/django_docker_engine
- Selenium-based tests to ensure our Visualization Container launches behave appropriately

### CODE REVIEW

File: refinery/tool_manager/admin.py [hunk 1]
```diff
@@ -1,7 +1,9 @@
 from django.contrib import admin
 
 from .models import (FileRelationship, GalaxyParameter, InputFile,
-                     OutputFile, Parameter, ToolDefinition)
+                     OutputFile, Parameter, ToolDefinition,
+                     VisualizationDefinition)
+
```
Comment (jkmarx) [inline refinery/tool_manager/admin.py]: extra line

File: refinery/tool_manager/management/commands/generate_tool_definitions.py [hunk 2]
```diff
@@ -1,14 +1,17 @@
...
-                      validate_tool_annotation,
-                      validate_workflow_step_annotation)
+from ...models import ToolDefinition
+from ...utils import (
+    ANNOTATION_ERROR_MESSAGE,
+    create_tool_definition,
+    get_visualization_annotations_list,
+    get_workflow_list,
+    validate_tool_annotation,
+    validate_workflow_step_annotation)
```
Comment (jkmarx) [inline refinery/tool_manager/management/commands/generate_tool_definitions.py]: fix ()

File: refinery/tool_manager/management/commands/generate_tool_definitions.py [hunk 3]
```diff
@@ -19,136 +22,198 @@ class Command(BaseCommand):
     main program; run the command
     """
 
-    def handle(self, **options):
+    option_list = BaseCommand.option_list + (
+        make_option(
+            '--visualizations',
```
Comment (jkmarx) [inline refinery/tool_manager/management/commands/generate_tool_definitions.py]: should visualization & workflow be plural?

File: refinery/tool_manager/models.py [hunk 4]
```diff
@@ -144,6 +153,78 @@ class ToolDefinition(models.Model):
...
 
+    def get_container_name(self):
+        """
+        Fetch the container_name of a VisualizationDefinition
+        :return: VisualizationDefinition.container_name if available,
+        otherwise `None`
+        """
+        try:
+            visualization_definition = self.get_visualization_definition()
+        except (VisualizationDefinition.DoesNotExist,
```
Comment (jkmarx) [inline refinery/tool_manager/models.py]: Aren't these separate issues?

File: refinery/tool_manager/test_data/visualization_LIST_hello_world.json [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    "output_files":[],
+    "file_relationship": {
+      "file_relationship": {},
+      "value_type": "LIST",
+      "name": "Flat list of N Samples",
+      "input_files": [
+        {
+          "allowed_filetypes": [{"name": "GFF"}],
+          "name": "Cool Input File",
+          "description": "Cool Input File Description"
```
Comment (jkmarx) [inline refinery/tool_manager/test_data/visualization_LIST_hello_world.json]: uuid? front end is expecting all input_files to have a uuid

File: refinery/tool_manager/test_data/visualization_LIST_igv.json [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+    "output_files":[],
+    "file_relationship": {
+      "file_relationship": {},
+      "value_type": "LIST",
+      "name": "Flat list of N Samples",
+      "input_files": [
+        {
+          "allowed_filetypes": [{"name": "SEG"}],
+          "name": "Cool Input File",
+          "description": "Cool Input File Description"
```
Comment (jkmarx) [inline refinery/tool_manager/test_data/visualization_LIST_igv.json]: uuid

File: refinery/tool_manager/tests.py [hunk 7]
```diff
@@ -82,6 +93,18 @@ def setUp(self):
 
     def test_tool_definitions_exist(self):
         self.assertEqual(ToolDefinition.objects.count(), 4)
+        self.assertEqual(
```
Comment (jkmarx) [inline refinery/tool_manager/tests.py]: Separate test, checking if the correct tool types are returned

File: refinery/tool_manager/views.py [hunk 8]
```diff
@@ -1,8 +1,15 @@
...
+
+from django.http import HttpResponseServerError
+
 from rest_framework.permissions import IsAuthenticated
-from rest_framework.viewsets import ModelViewSet
+from rest_framework.viewsets import ModelViewSet, GenericViewSet
+
+from .models import (ToolDefinition,
+                     VisualizationToolLaunch)
+from .serializers import (ToolDefinitionSerializer)
```
Comment (jkmarx) [inline refinery/tool_manager/views.py]: remove ()

File: refinery/tool_manager/views.py [hunk 9]
```diff
@@ -1,8 +1,15 @@
+import logging
+
+from django.http import HttpResponseServerError
+
 from rest_framework.permissions import IsAuthenticated
-from rest_framework.viewsets import ModelViewSet
+from rest_framework.viewsets import ModelViewSet, GenericViewSet
```
Comment (jkmarx) [inline refinery/tool_manager/views.py]: G before M

File: .travis.yml [hunk 10]
```diff
@@ -1,15 +1,19 @@
...
   apt:
     packages:
       - xvfb
+
 python:
   - 2.7
 services:
   - postgresql
   - neo4j
+  - docker
```
Comment (mccalluc) [inline .travis.yml]: Note: When I finish adding support for ECS in https://github.com/refinery-platform/django_docker_engine, the these tests could run either against that, or against the local docker engine like they do now.

File: refinery/tool_manager/schemas/ToolDefinition.json [hunk 11]
```diff
@@ -1,45 +1,86 @@
...
-    "file_relationship": {
-      "$ref": "FileRelationship.json#"
-    },
-    "output_files": {
-      "type": "array",
-      "items": {
-       "$ref": "OutputFile.json#"
-      },
-      "minItems": 1
+  "oneOf":[
```
Comment (mccalluc) [inline refinery/tool_manager/schemas/ToolDefinition.json]: It looks like the two definitions are mostly the same: Did you look at using `allOf` to extend definitions? https://spacetelescope.github.io/understanding-json-schema/structuring.html#extending

File: requirements.txt [hunk 12]
```diff
@@ -74,3 +74,5 @@ uuid==1.30
 watchdog==0.6.0
 xmltodict==0.9.2
 yolk==0.4.1
+
+git+https://github.com/mccalluc/django_docker_engine.git@v0.0.1
```
Comment (mccalluc) [inline requirements.txt]: This hits a redirect: It lives at https://github.com/refinery-platform/django_docker_engine now.

File: refinery/tool_manager/visualization_annotations/.gitignore [hunk 13]
```diff
@@ -0,0 +1 @@
+*.json
```
Comment (mccalluc) [inline refinery/tool_manager/visualization_annotations/.gitignore]: Can you explain this? (In person, if that's easier.) What's generating the files you want to ignore?

File: refinery/tool_manager/test_data/visualization_LIST_hello_world.json (hunk 5 above)
Comment (scottx611x) [inline refinery/tool_manager/test_data/visualization_LIST_hello_world.json]: This is just the ToolAnnotation, UUIDs are assigned when we create a ToolDefinition and put things into the Database

File: refinery/tool_manager/test_data/visualization_LIST_igv.json (hunk 6 above)
Comment (scottx611x) [inline refinery/tool_manager/test_data/visualization_LIST_igv.json]: This is just the ToolAnnotation, UUIDs are assigned when we create a ToolDefinition and put things into the Database

File: refinery/tool_manager/serializers.py [hunk 14]
```diff
@@ -31,7 +31,7 @@ def to_representation(self, obj):
         ret = super(ParameterSerializer, self).to_representation(obj)
 
         # Remove the `galaxy_workflow_step` field from the api response if it
-        # isn't available (Parameter objects for Visualization-based
+        # isn't available (Parameter objects for VisualizationDefinition-based
```
Comment (scottx611x) [inline refinery/tool_manager/serializers.py]: Change this back to `Visualization`

File: refinery/tool_manager/models.py (hunk 4 above)
Comment (scottx611x) [inline refinery/tool_manager/models.py]: What do you mean here?

File: refinery/tool_manager/management/commands/generate_tool_definitions.py (hunk 3 above)
Comment (scottx611x) [inline refinery/tool_manager/management/commands/generate_tool_definitions.py]: I think so, we can potentially load many wf and visualizations with these flags

File: refinery/selenium_testing/tests.py [hunk 15]
```diff
@@ -43,7 +43,7 @@ def setUp(self, site_login=True, initialize_guest=True,
             create_public_group()
 
     def tearDown(self):
-        self.browser.quit()
+        self.browser.close()
```
Comment (scottx611x) [inline refinery/selenium_testing/tests.py]: comment on `.quit()`s destructiveness

File: refinery/tool_manager/management/commands/generate_tool_definitions.py [hunk 16]
```diff
@@ -1,14 +1,17 @@
...
+from ...utils import (
+    ANNOTATION_ERROR_MESSAGE,
+    create_tool_definition,
+    get_visualization_annotations_list,
+    get_workflow_list,
+    validate_tool_annotation,
+    validate_workflow_step_annotation)
 
 
 class Command(BaseCommand):
```
Comment (scottx611x) [inline refinery/tool_manager/management/commands/generate_tool_definitions.py]: split into multiple mgmt commands and have helper methods in `utils.py`

File: refinery/tool_manager/migrations/0005_auto_20170412_1626.py [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+        ('tool_manager', '0004_visualizationcontainer'),
+    ]
+
+    operations = [
+        migrations.CreateModel(
+            name='VisualizationDefinition',
+            fields=[
+                ('tooldefinition_ptr', models.OneToOneField(parent_link=True, auto_created=True, primary_key=True, serialize=False, to='tool_manager.ToolDefinition')),
+                ('docker_image_name', models.CharField(max_length=255)),
+                ('container_name', models.CharField(max_length=150)),
```
Comment (scottx611x) [inline refinery/tool_manager/migrations/0005_auto_20170412_1626.py]: container_name shouldn't be unique. N containers for an image

File: refinery/tool_manager/models.py [hunk 18]
```diff
@@ -144,6 +153,78 @@ class ToolDefinition(models.Model):
...
+                VisualizationDefinition.MultipleObjectsReturned) as e:
+            logger.info(
+                "Couldn't properly fetch VisualizationDefinition with UUID: "
+                "{} {}".format(self.uuid, e)
+            )
+            raise
+
+
+class VisualizationDefinition(ToolDefinition):
```
Comment (scottx611x) [inline refinery/tool_manager/models.py]: Back into `ToolDefinition` these entines will be more similar if we move to an "All ECS" approach, but will simplify the API response stuff regardless.

`container_name` should be set in ToolLaunchConfiguration model (on the many side)

File: refinery/tool_manager/models.py [hunk 19]
```diff
@@ -185,3 +266,58 @@ def delete_input_files_and_file_relationships(sender, instance, *args,
...
     for file_relationship in file_relationships:
         file_relationship.delete()
+
+
+class ToolLaunchConfiguration(models.Model):
+    """
+    A ToolLaunchConfiguration is an abstract representation of the
+    information it will take to launch a Refinery Tool
+    """
+    file_relationships = models.TextField()
```
Comment (scottx611x) [inline refinery/tool_manager/models.py]: json.loads(TextField)

File: refinery/tool_manager/models.py [hunk 20]
```diff
@@ -185,3 +266,58 @@ def delete_input_files_and_file_relationships(sender, instance, *args,
...
+    parameters = models.TextField()
+    tool_definition = models.ForeignKey(ToolDefinition)
+    start_date = models.DateTimeField(auto_now_add=True)
+    # TODO: status. OwnableResource??? ????
+
+    class Meta:
+        abstract = True
```
Comment (scottx611x) [inline refinery/tool_manager/models.py]: Non abstract base?

File: refinery/tool_manager/models.py [hunk 21]
```diff
@@ -185,3 +266,58 @@ def delete_input_files_and_file_relationships(sender, instance, *args,
...
+        pass
+
+
+class VisualizationToolLaunch(ToolLaunchConfiguration):
+    def launch(self):
+        container = DockerContainerSpec(
+            image_name=self.tool_definition.get_docker_image_name(),
+            container_name=self.tool_definition.get_container_name(),
+            labels={self.tool_definition.uuid: "visualization_uuid"},
+            container_input_path="/var/www/data/input.json",
```
Comment (scottx611x) [inline refinery/tool_manager/models.py]: container_input_path non-hardcoded

File: refinery/tool_manager/tests.py [hunk 22]
```diff
@@ -485,3 +535,133 @@ def test_workflow_pair_not_enough_inputs(self):
                 workflow_annotation
             )
             self.assertEqual(ToolDefinition.objects.count(), 0)
+
+
+class ToolLaunchConfigurationTests(StaticLiveServerTestCase):
```
Comment (scottx611x) [inline refinery/tool_manager/tests.py]: Look into port conflicts (do theses live servers run in parallel?)

File: refinery/selenium_testing/tests.py (hunk 15 above)
Comment (hackdna) [inline refinery/selenium_testing/tests.py]: you should be able to use quit() if test cases are running sequentially

File: refinery/tool_manager/management/commands/generate_tool_definitions.py [hunk 23]
```diff
@@ -45,81 +45,152 @@ def handle(self, *args, **options):
         Workflows and Visualization Tools.
         """
 
+        if not options["visualizations"] and not options["workflows"]:
```
Comment (hackdna) [inline refinery/tool_manager/management/commands/generate_tool_definitions.py]: Replace with if... else

File: refinery/tool_manager/management/commands/generate_tool_definitions.py [hunk 24]
```diff
@@ -45,81 +45,152 @@ def handle(self, *args, **options):
...
-            generate_workflow_tool_definitions()
+            self.generate_workflow_tool_definitions()
 
+    def generate_visualization_tool_definitions(self):
+        """
+        Validate visualization annotation data, and try to create a
+        ToolDefinition if our validation rules pass.
+        """
+        self.stdout.write(
+            self.style.WARNING(
```
Comment (hackdna) [inline refinery/tool_manager/management/commands/generate_tool_definitions.py]: Why WARNING?

File: refinery/tool_manager/management/commands/generate_tool_definitions.py [hunk 25]
```diff
@@ -45,81 +45,152 @@ def handle(self, *args, **options):
...
+            ]:
+                self.stdout.write(
+                    self.style.NOTICE(
+                        "Skipping creation of `{0}` since ToolDefinition with "
+                        "name: `{0}` already exists.".format(
+                            visualization["name"]
+                        )
+                    )
+                )
+                pass
```
Comment (hackdna) [inline refinery/tool_manager/management/commands/generate_tool_definitions.py]: Why is pass necessary?

File: refinery/tool_manager/management/commands/generate_tool_definitions.py [hunk 26]
```diff
@@ -45,81 +45,152 @@ def handle(self, *args, **options):
...
-        try:
-            create_tool_definition(visualization)
-        except Exception as e:
-            raise CommandError(
-                "Creation of ToolDefinition failed. Database "
-                "rolled back to its state before this "
-                "ToolDefinition's attempted creation: {}".format(e)
-            )
+                self.stdout.write(
+                    self.style.WARNING(
```
Comment (hackdna) [inline refinery/tool_manager/management/commands/generate_tool_definitions.py]: Why WARNING?

File: refinery/tool_manager/migrations/0014_auto_20170419_1509.py [hunk 27]
```diff
@@ -0,0 +1,32 @@
...
+            field=models.CharField(default='', max_length=500, blank=True),
+            preserve_default=False,
+        ),
+        migrations.AlterField(
+            model_name='tooldefinition',
+            name='galaxy_workflow_id',
+            field=models.CharField(default='', max_length=250, blank=True),
+            preserve_default=False,
+        ),
+    ]
```
Comment (hackdna) [inline refinery/tool_manager/migrations/0014_auto_20170419_1509.py]: Why not squash all these migrations?

File: refinery/tool_manager/utils.py [hunk 28]
```diff
@@ -249,34 +292,86 @@ def validate_workflow_step_annotation(workflow_step_dictionary):
...
+            )
+        )
+
+
+def get_workflows():
     """
     Generate a list of available workflows from all currently available
     WorkflowEngines
-    :return: list of workflow dicts
+    :return: dict with keys == WorkflowEngine.uuid's and values == list of
```
Comment (hackdna) [inline refinery/tool_manager/utils.py]: Unclear and contradicts with the statement above.

File: .travis.yml (hunk 10 above)
Comment (hackdna) [inline .travis.yml]: Local engine

File: refinery/tool_manager/management/commands/generate_tool_definitions.py (hunk 24 above)
Comment (scottx611x) [inline refinery/tool_manager/management/commands/generate_tool_definitions.py]: I think i just missed this, it should be `NOTICE`

File: refinery/tool_manager/migrations/0014_auto_20170419_1509.py (hunk 27 above)
Comment (scottx611x) [inline refinery/tool_manager/migrations/0014_auto_20170419_1509.py]: I could, but this requires making sure that everyone has applied the prior migrations properly before I'm able to remove them. 

Django docs say:
```
You are encouraged to make migrations freely and not worry about how many you have; the migration code is optimized to deal with hundreds at a time without much slowdown. However, eventually you will want to move back from having several hundred migrations to just a few, and that’s where squashing comes in.
```

File: refinery/tool_manager/management/commands/generate_tool_definitions.py (hunk 26 above)
Comment (scottx611x) [inline refinery/tool_manager/management/commands/generate_tool_definitions.py]: I think i just missed this, it should be NOTICE

File: refinery/tool_manager/management/commands/generate_tool_definitions.py (hunk 25 above)
Comment (scottx611x) [inline refinery/tool_manager/management/commands/generate_tool_definitions.py]: It is not needed here

File: refinery/selenium_testing/tests.py [hunk 29]
```diff
@@ -43,7 +43,7 @@ def setUp(self, site_login=True, initialize_guest=True,
             create_public_group()
 
     def tearDown(self):
-        self.browser.close()
+        self.browser.quit()
```
Comment (scottx611x) [inline refinery/selenium_testing/tests.py]: add comment as to the use of `quit()`

File: refinery/tool_manager/management/commands/generate_tool_definitions.py (hunk 23 above)
Comment (scottx611x) [inline refinery/tool_manager/management/commands/generate_tool_definitions.py]: if elif else

File: refinery/tool_manager/management/commands/generate_tool_definitions.py [hunk 30]
```diff
@@ -128,25 +202,24 @@ def generate_workflow_tool_definitions():
...
+                    validate_workflow_step_annotation(step_annotation)
                 except RuntimeError as e:
                     raise CommandError(
                         "{} {}".format(ANNOTATION_ERROR_MESSAGE, e)
                     )
                 try:
                     parameters = step_annotation["parameters"]
                 except KeyError:
-                    # `parameters` aren't required for each workflow step
+                    # `parameters` aren't required
```
Comment (scottx611x) [inline refinery/tool_manager/management/commands/generate_tool_definitions.py]: new line?

File: refinery/tool_manager/management/commands/generate_tool_definitions.py [hunk 31]
```diff
@@ -162,16 +235,15 @@ def generate_workflow_tool_definitions():
...
+                            parameter["galaxy_workflow_step"] = int(step_index)
                             workflow["annotation"]["parameters"].append(
                                 parameter
                             )
                 try:
                     output_files = step_annotation["output_files"]
                 except KeyError:
-                    # `output_files` aren't required for each workflow step
+                    # `output_files` aren't required
+                    # for each workflow step
```
Comment (scottx611x) [inline refinery/tool_manager/management/commands/generate_tool_definitions.py]: new line ?

File: refinery/tool_manager/models.py [hunk 32]
```diff
@@ -268,56 +207,72 @@ def delete_input_files_and_file_relationships(sender, instance, *args,
...
+                )
+
+        if self.get_tool_type() == ToolDefinition.WORKFLOW:
+            raise NotImplementedError
+
+    def get_relative_container_url(self):
+        """
+        Construct & return the relative url of our Tool's container
+        """
+        return "/api/v2/docker/{}".format(self.container_name)
```
Comment (scottx611x) [inline refinery/tool_manager/models.py]: /api/v2/ seems wrong here. Just specifying `/docker` would be better

File: refinery/tool_manager/tests.py [hunk 33]
```diff
@@ -44,19 +44,31 @@ def setUp(self):
...
         # Make some sample data
         with open(
                 "{}/visualization_LIST_hello_world.json".format(TEST_DATA_PATH)
         ) as f:
-            create_tool_definition(json.loads(f.read()))
+            tool_annotation = json.loads(f.read())
+            create_tool_definition(tool_annotation)
         with open("{}/workflow_LIST.json".format(TEST_DATA_PATH)) as f:
-            create_tool_definition(json.loads(f.read()))
+            tool_annotation = json.loads(f.read())
```
Comment (scottx611x) [inline refinery/tool_manager/tests.py]: could just be `uuid.uuid4()` in tests

File: refinery/tool_manager/tests.py [hunk 34]
```diff
@@ -134,23 +146,22 @@ def test_for_proper_parameters_in_response(self):
             for parameter in tool_definition["parameters"]:
                 if tool_definition["tool_type"] == ToolDefinition.WORKFLOW:
                     self.assertIn("galaxy_workflow_step", parameter.keys())
-                elif tool_definition["tool_type"] == \
-                        ToolDefinition.VISUALIZATION:
+                elif (tool_definition["tool_type"] ==
```
Comment (scottx611x) [inline refinery/tool_manager/tests.py]: this test can go away

File: refinery/tool_manager/tests.py [hunk 35]
```diff
@@ -557,21 +720,21 @@ def setUp(self):
         self.user = User.objects.create_user(self.username, '', self.password)
         self.container_name = ""
         self.factory = APIRequestFactory()
-        self.view = ToolLaunchConfigurationViewSet.as_view({'post': 'launch'})
+        self.view = ToolViewSet.as_view({'post': 'create'})
         self.url_root = '/api/v2/tools'
 
     def tearDown(self):
-        self.browser.close()
+        self.browser.quit()
```
Comment (scottx611x) [inline refinery/tool_manager/tests.py]: make comment on quit()

File: refinery/tool_manager/tests.py [hunk 36]
```diff
@@ -557,21 +720,21 @@ def setUp(self):
...
         self.display.stop()
         try:
-            visualization_definition = VisualizationDefinition.objects.get(
+            tool_launch = Tool.objects.get(
```
Comment (scottx611x) [inline refinery/tool_manager/tests.py]: no get() needed in tearDown()

File: refinery/tool_manager/tests.py [hunk 37]
```diff
@@ -604,10 +765,19 @@ def test_visualization_container_launch_and_access_hello_world(self):
             )
             force_authenticate(self.post_request, self.user)
             self.post_response = self.view(self.post_request)
+
+            tool_launch = Tool.objects.all()[0]
```
Comment (scottx611x) [inline refinery/tool_manager/tests.py]: use tool_definition.uuid to get()

File: refinery/tool_manager/urls.py [hunk 38]
```diff
@@ -1,15 +1,11 @@
...
 
 # DRF url routing
 tool_manager_router = DefaultRouter()
 tool_manager_router.register(r'tools/definitions', ToolDefinitionsViewSet)
-tool_manager_router.register(
-    r'tools',
-    ToolLaunchConfigurationViewSet,
-    "ToolLaunch"
-)
+tool_manager_router.register(r'tools', ToolViewSet)
```
Comment (scottx611x) [inline refinery/tool_manager/urls.py]: to -> ToolsViewset

File: refinery/tool_manager/urls.py [hunk 39]
```diff
@@ -1,15 +1,11 @@
 from django.conf.urls import url, include
 from rest_framework.routers import DefaultRouter
-from .views import ToolDefinitionsViewSet, ToolLaunchConfigurationViewSet
+from .views import ToolDefinitionsViewSet, ToolViewSet
 
 # DRF url routing
 tool_manager_router = DefaultRouter()
 tool_manager_router.register(r'tools/definitions', ToolDefinitionsViewSet)
```
Comment (scottx611x) [inline refinery/tool_manager/urls.py]: r'tools/definitions' -> r'tool_definitions'

File: refinery/tool_manager/views.py [hunk 40]
```diff
@@ -22,16 +23,24 @@ class ToolDefinitionsViewSet(ModelViewSet):
...
+    """API endpoint that allows for Tools to be fetched and launched"""
+
+    queryset = Tool.objects.all()
+    serializer_class = ToolSerializer
+    lookup_field = 'uuid'
+    http_method_names = ['get', 'post']
     permission_classes = [IsAuthenticated]
 
-    def launch(self, request, *args, **kwargs):
+    @transaction.atomic
```
Comment (scottx611x) [inline refinery/tool_manager/views.py]: use `with transaction.atomic()` instead of decorating entire view

File: refinery/tool_manager/views.py [hunk 41]
```diff
@@ -45,22 +54,35 @@ def launch(self, request, *args, **kwargs):
...
-                        "parameters"
-                    ],
-                    file_relationships=tool_launch_data[
-                       "file_relationships"
-                    ]
+                    parameters=tool_launch_data["parameters"],
+                    file_relationships=tool_launch_data["file_relationships"]
+                )
 
+                # Create a unique container name that adheres to dockers
```
Comment (scottx611x) [inline refinery/tool_manager/views.py]: "docker's"

File: refinery/tool_manager/views.py [hunk 42]
```diff
@@ -45,22 +54,35 @@ def launch(self, request, *args, **kwargs):
...
+                # specifications
+                visualization_tool.container_name = "{}-{}".format(
+                    visualization_tool.name.replace(" ", ""),
+                    visualization_tool.uuid
                 )
+
+                visualization_tool.set_owner(request.user)
+                visualization_tool.save()
+
                 return visualization_tool.launch()
```
Comment (scottx611x) [inline refinery/tool_manager/views.py]: just return url

**Pull request merged**